### PR TITLE
feat: track epoch transition steps time in metrics

### PR DESCRIPTION
**Pull request by twoeths**
**Motivation**

We have unstable epoch transitions and we have no idea which steps take the most time

Screenshot 2023-11-30 at 11 42 29

**Description**

- Track every step of epoch transition in metrics

part of #6112
part of #6063
part of #5409

### CODE REVIEW

**Comment by dapplion** [comment]
Definitely not necessary to track every single step. Some functions are extremely cheap to run, just track the ones that do complex expensive logic, look at the benchmarks for reference

File: packages/beacon-node/src/metrics/metrics/lodestar.ts [hunk 1]
```diff
@@ -286,6 +286,56 @@ export function createLodestarMetrics(
...
+      name: "lodestar_stfn_epoch_transition_before_process_epoch_seconds",
+      help: "Time to call beforeProcessEpoch before process a single epoch transition in seconds",
+      buckets: [0.01, 0.05, 0.1, 0.2, 0.5, 0.75, 1],
+    }),
+    epochTransitionAfterProcessEpochTime: register.histogram({
+      name: "lodestar_stfn_epoch_transition_after_process_epoch_seconds",
+      help: "Time to call afterProcessEpoch after process a single epoch transition in seconds",
+      buckets: [0.01, 0.05, 0.1, 0.2, 0.5, 0.75, 1],
+    }),
+    epochTransitionJustificationAndFinalizationTime: register.histogram({
```
Comment (wemeetagain) [inline packages/beacon-node/src/metrics/metrics/lodestar.ts]: would it be worth defining all steps from a single histogram metric with a "step" label?

Comment (twoeths) [reply packages/beacon-node/src/metrics/metrics/lodestar.ts]: oh yes, should be less code with that 👍

File: packages/state-transition/src/epoch/index.ts [hunk 2]
```diff
@@ -59,19 +65,31 @@ export function processEpoch(fork: ForkSeq, state: CachedBeaconStateAllForks, ca
     throw new Error("Lodestar does not support this network, parameters don't fit number value inside state.slashings");
   }
 
+  let timer = metrics?.epochTransitionStepTime.startTimer({step: "processJustificationAndFinalization"});
```
Comment (dapplion) [inline packages/state-transition/src/epoch/index.ts]: I don't think it's necessary to use a let here, just declare individual variables. You only need to register labels once

**Review by wemeetagain — APPROVED** [review]

File: packages/state-transition/src/epoch/index.ts [hunk 3]
```diff
@@ -59,19 +65,53 @@ export function processEpoch(fork: ForkSeq, state: CachedBeaconStateAllForks, ca
     throw new Error("Lodestar does not support this network, parameters don't fit number value inside state.slashings");
   }
 
-  processJustificationAndFinalization(state, cache);
+  {
+    const timer = metrics?.epochTransitionStepTime.startTimer({
```
Comment (nflaig) [inline packages/state-transition/src/epoch/index.ts:69]: would it make sense to have a metric utility function `withTimer` which just does the instrumentation? could refactor in a separate PR

Comment (wemeetagain) [reply packages/state-transition/src/epoch/index.ts:69]: yeah, that would be nice

**Review by wemeetagain — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.13.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.13.0) :tada:

**Pull request merged**